src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs

```diff
@@ -46,19 +46,32 @@ public void AddAllTo(HashSet<Checksum> checksums)
     [PerformanceSensitive("https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1333566", AllowGenericEnumeration = false)]
     internal static async Task FindAsync<TState>(
         TextDocumentStates<TState> documentStates,
+        DocumentId? hintDocument,
         HashSet<Checksum> searchingChecksumsLeft,
         Dictionary<Checksum, object> result,
         CancellationToken cancellationToken) where TState : TextDocumentState
     {
-        foreach (var (_, state) in documentStates.States)
+        if (hintDocument != null)
```

Comment:
tweaked to go to O(1) lookup, instead of the linear walk when the hint is present.